ckan/migration/migrate_package_activity.py

```diff
@@ -0,0 +1,261 @@
+        #  'timestamp': '2018-04-20T16:11:45.363097',
+        #  'user_id': u'724273ac-a5dc-482e-add4-adaf1871f8cb'}
+        print(u'  activity {}/{} {}'.format(
+              i + 1, num_activities, activity[u'timestamp']))
+
+        # we need activity.data and using the ORM is the fastest
+        activity_obj = model.Session.query(model.Activity).get(activity[u'id'])
+        if u'resources' in activity_obj.data.get(u'package', {}):
+            print(u'    activity has full dataset already recorded'
+                  ' - no action')
+            continue
+
+        # get the dataset as it was at this revision
+        context[u'revision_id'] = activity[u'revision_id']
+        # call package_show just as we do in package.py:activity_stream_item(),
+        # only with a revision_id
+        try:
+            dataset = logic.get_action(u'package_show')(
+                context,
+                {u'id': activity[u'object_id'], u'include_tracking': False})
+        except logic.NotFound as exc:
+            print(u'    Revision missing! Skipping this version '
+                  '(revision_id={})'.format(activity[u'revision_id']))
+            errors['Revision missing'] += 1
+            # We shouldn't leave the activity.data['package'] with missing
+            # resources, extras & tags, which could cause the package_read
+            # template to raise an exception, when user clicks "View this
+            # version". Instead we pare it down to use a title, and forgo
+            # viewing it.
+            try:
+                dataset = {u'title': activity_obj.data['package']['title']}
+            except KeyError:
+                dataset = None
+
+        # get rid of revision_timestamp, which wouldn't be there if saved by
+        # during activity_stream_item() - something to do with not specifying
+        # revision_id.
+        if u'revision_timestamp' in (dataset.get(u'organization') or {}):
+            del dataset[u'organization'][u'revision_timestamp']
+        for res in dataset.get(u'resources', []):
+            if u'revision_timestamp' in res:
+                del res[u'revision_timestamp']
+
+        actor = model.Session.query(model.User).get(activity[u'user_id'])
+        actor_name = actor.name if actor else activity[u'user_id']
+
+        # add the data to the Activity, just as we do in activity_stream_item()
+        data = {
+            u'package': dataset,
+            u'actor': actor_name,
+        }
+        activity_obj.data = data
+        # print '    {} dataset {}'.format(actor_name, repr(dataset))
+    if model.Session.dirty:
+        model.Session.commit()
+        print(u'  saved')
+    print(u'  This package\'s {} activities are migrated'.format(
+        len(package_activity_stream)))
+
+
+def wipe_activity_detail(delete_activity_detail):
+    from ckan import model
+    num_activity_detail_rows = \
+        model.Session.execute(u'SELECT count(*) FROM "activity_detail";') \
```

Comment:
`SELECT count(*) from (SELECT 1 FROM "activity_detail" LIMIT 1) as t` to make this query faster

Reply:
cool, changed